From 26e4f2ad9112cc9c38217a4ab75777064d6fe5cd Mon Sep 17 00:00:00 2001
From: yzha108 <yanli.l.zhang@intel.com>
Date: Tue, 28 Aug 2018 01:32:11 +0800
Subject: [PATCH 41/54] media: intel-ipu4: ar0231: add test pattern ioctl.

Message for Open Source:
Add test pattern ioctl support

Message for Internal:
[Issue/Feature]
Add test pattern ioctl support

[Root Cause/Changes]
Add test pattern ioctl support

Change-Id: I8b68634bb789ff5f13791fb30d9c7362d82715da
TrackedOn: #JIIAP-586
Signed-off-by: yzha108 <yanli.l.zhang@intel.com>
(cherry picked from commit bbf0310291c2e38abdebec251f389716291eba10)
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 .../i2c/crlmodule/crl_ar0231at_configuration.h     | 33 ++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h b/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
index ccca157..0daeb2f 100644
--- a/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
+++ b/drivers/media/i2c/crlmodule/crl_ar0231at_configuration.h
@@ -2060,6 +2060,22 @@ struct crl_csi_data_fmt ar0231at_crl_csi_data_fmt[] = {
 	},
 };
 
+static struct crl_dynamic_register_access ar0231at_test_pattern_regs[] = {
+	{
+		.address = 0x3070,
+		.len = CRL_REG_LEN_16BIT,
+		.ops_items = 0,
+		.ops = 0,
+		.mask = 0xffff,
+	},
+};
+
+static const char * const ar0231at_test_patterns[] = {
+	"Disabled",
+	"Solid Color",
+	"100% Vertical Color Bar",
+};
+
 struct crl_v4l2_ctrl ar0231at_v4l2_ctrls[] = {
 	{
 		.sd_type = CRL_SUBDEV_TYPE_BINNER,
@@ -2311,6 +2327,23 @@ struct crl_v4l2_ctrl ar0231at_v4l2_ctrls[] = {
 		.dep_ctrls = 0,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
+	{
+		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
+		.op_type = CRL_V4L2_CTRL_SET_OP,
+		.context = SENSOR_POWERED_ON,
+		.ctrl_id = V4L2_CID_TEST_PATTERN,
+		.name = "V4L2_CID_TEST_PATTERN",
+		.type = CRL_V4L2_CTRL_TYPE_MENU_ITEMS,
+		.data.v4l2_menu_items.menu = ar0231at_test_patterns,
+		.data.v4l2_menu_items.size = ARRAY_SIZE(ar0231at_test_patterns)-1,
+		.flags = V4L2_CTRL_FLAG_UPDATE,
+		.impact = CRL_IMPACTS_NO_IMPACT,
+		.ctrl = 0,
+		.regs_items = ARRAY_SIZE(ar0231at_test_pattern_regs),
+		.regs = ar0231at_test_pattern_regs,
+		.dep_items = 0,
+		.dep_ctrls = 0,
+	},
 };
 
 struct crl_sensor_detect_config ar0231at_sensor_detect_regset[] = {
-- 
1.9.1

